rfcs/shared/build-system/ssr-testing.md

```diff
@@ -0,0 +1,240 @@
+
+The proposal is to implement a new application that will perform tests. We could use Next.js or other framework to perform rendering, but it's too heavy for our tests.
+
+```mermaid
+flowchart TB
+    subgraph Build
+    b1(Get all existing stories and generate an application)
+
+      subgraph Build an application
+        direction BT
+        b21(Build CommonJS output)
+        b22(Build ESM output)
+      end
+
+    b3(Prepare assets)
+
+    b1-->b21
+    b1-->b22
+    b21-->b3
+    b22-->b3
+    end
+
+    b3-->t1
+
+
+    subgraph Test
+      t1(Open a browser)
+      t2(Open a page and run JS)
+      t3(Ensure that console is empty)
+
+      t1-->t2
+      t2-->t3
+    end
+```
+
+### Get all existing stories and generate an application
+
+We have the source of truth in `packages/react-components/.storybook/main.utils`, there is a function `getVnextStories` that returns globs to find files with stories.
+
+<details>
+  <summary>`getStories.js`</summary>
+
+```js
+const { getVnextStories } = require('./packages/react-components/.storybook/main.utils');
+
+const stories = getVnextStories()
+  .map(g => {
+    return glob.sync(g, {
+      cwd: path.resolve(__dirname, 'packages/react-components/.storybook'),
+    });
+  })
+  .flatMap(i => i)
+  .map(f => path.resolve(__dirname, 'packages/react-components/.storybook', f))
+  .filter(f => fs.readFileSync(f, 'utf8').includes('export { Default } from'))
+  .map(f =>
+    fs
+      .readFileSync(f, 'utf8')
+      .split('\n')
+      .filter(l => l.includes('export {'))
+      .map(l =>
+        l
+          .replace('export {', 'import {')
+          .replace(/\'(.\/)(.+)\';/g, (...a) => {
+            const importName = a[2];
+            return `"./${path.relative(__dirname, path.dirname(f))}/${importName}";`;
+          })
+          .replace(/import { (.+) }/g, (...a) => {
+            const base = path.basename(f).split('.')[0];
+            return `import { ${a[1]} as ${base}${a[1]} }`;
+          }),
+      ),
+  )
+  .flatMap(f => f);
+```
+
+</details>
+
+> **Notes**:
+>
+> - For the final implementation it probably will use AST instead of regular expressions
+> - There is no sense to use Storybook API as we have strict structure for stories (a single story per a file)
+
+This ugly snippet generates following:
+
+```ts
+import { Default as DividerDefault } from './packages/react-divider/src/stories/DividerDefault.stories';
+import { Vertical as DividerVertical } from './packages/react-divider/src/stories/DividerVertical.stories';
+// etc.
+```
+
+With additional tweaks we can generate the following `App` component:
+
+```tsx
+// stories.tsx
+import { Truncate as TextTruncate } from './packages/react-text/src/stories/TextTruncate.stories';
+import { Font as TextFont } from './packages/react-text/src/stories/TextFont.stories';
+
+export const Stories = () => (
+  <SSRProvider>
+    <FluentProvider theme={teamsLightTheme}>
+      <TextTruncate />
```

Comment:
What is a report for an SSR test going to look like?

In the example all the stories for a component are in a single test but it's possible that only one story triggers an SSR failure. If all the stories are grouped into a single test it could be frustrating to debug a failure.